compactor/src/compact.rs

```diff
@@ -1231,6 +1270,48 @@ impl Compactor {
     }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+struct OverlappedGroups {
+    // Groups that contain overlapped files but the groups do not overlapped
+    // with other groups
+    internal_overlapped_groups: Vec<Vec<ParquetFile>>,
+
+    // Groups that overlap with at least one of other groups
+    external_overlapped_groups: Vec<Vec<ParquetFile>>,
+}
+
+impl OverlappedGroups {
+    pub fn new(len: usize) -> Self {
+        Self {
+            internal_overlapped_groups: Vec::with_capacity(len),
+            external_overlapped_groups: Vec::with_capacity(len),
+        }
+    }
+
+    pub fn add_internal_overalapped_group(&mut self, group: Vec<ParquetFile>) {
+        self.internal_overlapped_groups.push(group);
+    }
+
+    pub fn add_external_overalapped_group(&mut self, group: Vec<ParquetFile>) {
+        self.external_overlapped_groups.push(group);
+    }
+
+    pub fn groups_with_min_time_and_size(&mut self) -> Vec<GroupWithMinTimeAndSize> {
+        let mut groups = Vec::with_capacity(
+            self.internal_overlapped_groups.len() + self.external_overlapped_groups.len(),
+        );
+
+        for group in &self.internal_overlapped_groups {
+            groups.push(GroupWithMinTimeAndSize::new(group.to_vec(), false));
+        }
+        for group in &self.external_overlapped_groups {
+            groups.push(GroupWithMinTimeAndSize::new(group.to_vec(), true));
+        }
+
+        groups
+    }
```

Comment:
Calling `to_vec()` will clone all the parquet files (which could be a lot!).

We never re-use this `OverlappedGroups` after we have called `groups_with_min_time_and_size()` so we can _move_ the vecs, and not have to clone them. 

We do this by changing the fn sig to take `self` (not `&mut self`) and moving the vecs:

```suggestion
    pub fn groups_with_min_time_and_size(self) -> Vec<GroupWithMinTimeAndSize> {
        let mut groups = Vec::with_capacity(
            self.internal_overlapped_groups.len() + self.external_overlapped_groups.len(),
        );

        for group in self.internal_overlapped_groups {
            groups.push(GroupWithMinTimeAndSize::new(group, false));
        }
        for group in self.external_overlapped_groups {
            groups.push(GroupWithMinTimeAndSize::new(group, true));
        }

        groups
    }
```